npm_helpers.py

```diff
@@ -0,0 +1,103 @@
+    with temporary_package_json(root_dir, dependencies):
+        try:
+            # Install all the dependencies
+            # TODO: Should NPM itself have resource limits?
+            # TODO: Should we prune, or delete node_modules?
+            subprocess.check_output([settings.NPM_BINARY, "prune"], stderr=subprocess.STDOUT, cwd=root_dir)
+            if dependencies:
+                subprocess.check_output([settings.NPM_BINARY, "install", "--ignore-scripts"], stderr=subprocess.STDOUT, cwd=root_dir)
+        except subprocess.CalledProcessError:
+            # Setting the error message to e.output here would let the user see their lovely NPM error output.
+            raise NPMInstallError("One or more of your dependencies cannot be installed. Please check that the names and versions for all of your dependencies are valid.")
+
+
+def get_package_metadata(root_dir):
+    """ Given a directory with a node_modules directory full of pebble modules, get all of their messageKeys and resource type/name information
+    :param root_dir:
+    :return: (array of resources, array of (type, name) tuples
+    """
+    resources = []
+    messagekeys = []
+    try:
+        for package_path in glob.glob(os.path.join(root_dir, 'node_modules', '*', 'package.json')):
+            with open(package_path, 'r') as f:
+                data = json.load(f)
+                messagekeys.extend(data['pebble'].get('messageKeys', []))
+                resources.extend(Resource(r['type'], r['name']) for r in data['pebble'].get('resources', {}).get('media', []))
+        return resources, messagekeys
+    except Exception:
+        raise NPMInstallError("One or more of your dependencies is not a valid pebble library.")
+
+
+def extract_library_headers(root_dir):
+    """ Given a directory with a node_modules directory full of Pebble modules, extract all of their header files into ./librares/include/<module_name>
+    :param root_dir: base directory
+    """
+    # Make the libraries directory if it does not exist
+    libs_path = os.path.join(root_dir, 'libraries')
+    if os.path.isdir(libs_path):
+        shutil.rmtree(libs_path)
+
+    os.mkdir(libs_path)
+    # Look for C modules with dist.zip files
+    for zip_path in glob.glob(os.path.join(root_dir, 'node_modules', '*', 'dist.zip')):
+        try:
+            # Construct the expected path to the library's headers based on its name
+            includes_path = os.path.join('include', os.path.basename(os.path.dirname(zip_path)))
+            with zipfile.ZipFile(zip_path) as z:
+                # Extract any header files which are inside 'include/<module_name>'
+                for zip_entry in z.infolist():
+                    if zip_entry.filename.startswith(includes_path) and zip_entry.filename.endswith('.h'):
+                        z.extract(zip_entry, libs_path)
+        except Exception:
+            raise NPMInstallError("One or more of your dependencies is not a valid pebble library.")
```

Comment:
Given that in this case my try/catch is only run if dist.zip is actually present, I feel like it might be appropriate to keep this one since it will likely only break for pebble objects which have dodgy dist.zips